_posts/2024/2024-02-06-about-caching-in-eo.md

```diff
@@ -0,0 +1,241 @@
+The [`ccache` hash](https://ccache.dev/manual/4.8.2.html#_common_hashed_information) is
+based on:
+* the file contents
+* the current directory of the file
+* the name of the compiler
+* the compiler’s size and modification time
+* extensions used by the compiler.
+
+Moreover, `ccache` has two types of the hashing:
+1) `Direct mode` - the hash is generated based on the source code only.
+When using this mode, the user must ensure that the external libraries used in a project have not changed.
+Otherwise, the project will fail to build, resulting in errors.
+2) `Preprocessor mode` - hash is generated based on the `.cpp` file received after the preprocessor step.
+
+
+`Sccache` is similar in purpose to `ccache` but provides more functionality.
+`Sccache` allows to store cached files not only locally, but also in a cloud data storage.
+And `sccache` supports a wider range of languages, while `ccache` focuses on caching C and C++ compiler.
+
+
+### Gradle
+[Gradle](https://gradle.org) builds projects using a
+[task graph](https://docs.gradle.org/current/userguide/build_lifecycle.html) that allows for synchronous execution 
+of certain tasks.
+`Gradle` employs
+[Incremental build](https://docs.gradle.org/current/userguide/incremental_build.html#sec:how_does_it_work),
+to speed up project builds.
+For an incremental build to work, the tasks used to build the project must have specified
+input and output files.
+The provided code snippet demonstrates the implementation of a custom task in Gradle,
+showcasing how inputs and outputs are specified to enable `Incremental build`:
+```
+task myTask {
+    inputs.file 'src/main/java/MyTask.somebody' // Specify the input file
+    outputs.file 'build/classes/java/main/MyTask.somebody' // Specify the output file
+    
+    doLast {
+        // Task actions go here
+        // This code will only be executed if the inputs or outputs have changed
+    }
+}
+```
+
+
+To understand how `Incremental build` works, consider the following steps:
+1) Before executing a task, `Gradle` takes a 
+   [fingerprint](https://en.wikipedia.org/wiki/Fingerprint_(computing))
+   of the path and contents of the inputs files and saves it.
+2) Then `Gradle` executes the task and saves a fingerprint of the path and contents of the output files.
+3) Then, when Gradle starts a project build again, it generates a new fingerprint for the same files.
+   If the new fingerprint has not changed, Gradle can safely skip this task. 
+   In the opposite case, the task needs to perform an action and to rewrite outputs.
+   The fingerprint is considered current if the last modification time 
+   and the size of the source files have not changed.
+
+
+In addition to `Incremental build`, `Gradle` also stores fingerprints of previous builds, enabling quick project builds,
+for example when switching from one branch to another. This feature is known as 
```

Comment:
@Yanich96 Please, specify which branch do you mean. Is it a "git" branch?

Reply:
@volodya-lombrozo yes, "git" branch